ui-v2/app/controllers/dc/acls/tokens/edit.js

```diff
@@ -0,0 +1,72 @@
+import Controller from '@ember/controller';
+import { inject as service } from '@ember/service';
+import { get, set } from '@ember/object';
+export default Controller.extend({
+  dom: service('dom'),
+  builder: service('form'),
+  isScoped: false,
+  init: function() {
+    this._super(...arguments);
+    this.form = get(this, 'builder').form('token');
+  },
+  setProperties: function(model) {
+    // essentially this replaces the data with changesets
+    this._super(
+      Object.keys(model).reduce((prev, key, i) => {
+        switch (key) {
+          case 'item':
+            prev[key] = this.form.setData(prev[key]).getData();
+            break;
+          case 'policy':
+            prev[key] = this.form
+              .form(key)
+              .setData(prev[key])
+              .getData();
+            break;
+        }
+        return prev;
+      }, model)
+    );
+  },
+  actions: {
+    sendClearPolicy: function(item) {
+      set(this, 'isScoped', false);
+      this.send('clearPolicy', item);
+    },
+    refreshCodeEditor: function(selector, parent) {
+      if (parent.target) {
+        parent = undefined;
+      }
+      get(this, 'dom')
+        .component(selector, parent)
+        .didAppear();
```

Comment:
Ah this was brought up previously here:

https://github.com/hashicorp/consul/pull/4190#discussion_r193270223

Same comments apply here, plus its what I'm already doing in the rest of the app, so I want to stay consistent.

I did spend quite a lot of time looking into the best way to do this within ember, most ember-y people out there do similar things (i.e. don't use what you would call traditional ember methods).

Re: data down actions up, it is being called on an event listener, it just calls a method instead of passing a property back down. Really no state is changing so that kind of make sense. It's only needed because when codemirror is initially rendered it has a `display :none` so it doesn't render properly on 'didInsertElement` this was just my 'ember-y' sounding lifecycle hook for when the element is actually displayed.

Reply:
Hey,

So the example you've given here is exactly what I want to avoid.

We've spoken about this before in a previous PR and offline following that, I'm not sure if you remember but I shared a few links with you, one was from Sam Selikoff (who I believe maintains mirage?)

http://www.samselikoff.com/blog/getting-ember-components-to-respond-to-actions/

> Using data-binding we could introduce a dummy property, have the street view component change that property, and have the fullscreen component respond to changes on that property; but this feels hackish, and the dummy variable wouldn't really represent anything meaningful in our application.

https://emberigniter.com/communication-between-distant-components/

This one ^ isn't _that_ similar to what I want to do here, and even this EventBus example is too coupled for me for my specific case, but:

> On the other hand, “Data down actions up” ain’t no religion. It’s important to know our tools.

I did see another article recently (dated a few weeks ago) that also discussed the same thing and had a similar solution, if I can find it again I'll post it here. I'll see if I can find that old PR were we discussed this also.

Reply:
Hey,

I found the PR where we spoke about this previously:

https://github.com/hashicorp/consul/pull/4190#discussion_r193270223

I also found the more recent article related to this last night:

https://medium.com/freshworks-engineering-blog/ember-accessing-child-component-functions-in-parent-f7b0fc1ad7fa

^ This one was featured in the Ember Times 3-4 weeks ago.

John